src/ScriptExport/ScriptInfo.hs

```diff
@@ -19,15 +33,213 @@ module ScriptExport.ScriptInfo (
+import Plutarch.Api.V2 (PMintingPolicy, PStakeValidator, PValidator, mkMintingPolicy, mkStakeValidator, mkValidator)
+import PlutusLedgerApi.V2 (MintingPolicy (getMintingPolicy), Script, StakeValidator (getStakeValidator), Validator (getValidator))
+import Ply (ScriptRole, TypedScript, TypedScriptEnvelope)
+import Ply.Core.TypedReader (TypedReader, mkTypedScript)
+
+-- | @since 2.0.0
+data LinkerError
+  = VersionMismatch
+  | ScriptNotFound Text
+  | ScriptTypeMismatch Text
+  | Other Text
+  deriving stock
+    ( -- | @since 2.0.0
+      Show
+    , -- | @since 2.0.0
+      Eq
+    , -- | @since 2.0.0
+      GHC.Generic
+    )
+  deriving anyclass
+    ( -- | @since 2.0.0
+      Aeson.ToJSON
+    , -- | @since 2.0.0
+      Aeson.FromJSON
+    )
+
+{- | Type for linker that makes ready-to-go script from parameter and
+     raw script.
+
+ @since 2.0.0
+-}
+newtype Linker param a
+  = Linker (ReaderT (RawScriptExport, param) (Except LinkerError) a)
+  deriving
+    ( -- | @since 2.0.0
+      Functor
+    , -- | @since 2.0.0
+      Applicative
+    , -- | @since 2.0.0
+      Monad
+    , -- | @since 2.0.0
+      MonadError LinkerError
+    , -- | @since 2.0.0
+      MonadReader (RawScriptExport, param)
+    )
+    via (ReaderT (RawScriptExport, param) (Except LinkerError))
+
+{- | Run linker with parameter and raw scripts.
+
+ @since 2.0.0
+-}
+runLinker ::
+  forall (param :: Type) (a :: Type).
+  Linker param (ScriptExport a) ->
+  RawScriptExport ->
+  param ->
+  Either LinkerError (ScriptExport a)
+runLinker (Linker linker) rs p =
+  runExcept $ runReaderT linker (rs, p)
+
+{- | Type for holding parameterized scripts. This type represents the
+     raw scripts which needs to be linked in order to be deployed.
+
+ @since 2.0.0
+-}
+data RawScriptExport = RawScriptExport
+  { version :: String
+  , scripts :: Map Text TypedScriptEnvelope
+  }
+  deriving stock
+    ( -- | @since 2.0.0
+      Show
+    , -- | @since 2.0.0
+      Eq
+    , -- | @since 2.0.0
+      GHC.Generic
+    )
+  deriving anyclass
+    ( -- | @since 2.0.0
+      Aeson.ToJSON
+    , -- | @since 2.0.0
+      Aeson.FromJSON
+    )
+
+-- | @since 2.0.0
+fetchTS ::
+  forall (rl :: ScriptRole) (params :: [Type]) (lparam :: Type).
+  TypedReader rl params =>
+  Text ->
+  Linker lparam (TypedScript rl params)
+fetchTS t = do
+  (scr, _) <- ask
+  case view #scripts scr !? t of
+    Just scr' ->
+      case mkTypedScript @rl @params scr' of
+        Right x -> return x
+        Left err -> throwError $ ScriptTypeMismatch $ pack $ show err
+    Nothing -> throwError $ ScriptNotFound t
+
+-- | @since 2.0.0
+exportParam :: forall (lparam :: Type). Linker lparam lparam
+exportParam = asks snd
+
+-- | @since 2.0.0
+exportVersion :: forall (lparam :: Type). Linker lparam String
+exportVersion = asks (view #version . fst)
+
+{- | Type for holding ready-to-go scripts.
```

Comment:
Likewise, let's define what "ready-to-go" here means -- e.g., `PData -> PScriptContext -> POpaque` for minting policies or staking validators, and `PData -> PData -> PScriptContext -> POpaque` for regular validators